### PR TITLE
Fix several issues in draw_mt tutorial

### PR DESCRIPTION
**Pull request by Axel-Naumann**

### CODE REVIEW

**Comment by linev** [comment]
@Axel-Naumann 

Can you apply all these changes after #2603?
Most probably, you will need to rebase your code - there are some changes in all these files.

File: net/http/src/THttpWSEngine.cxx [hunk 1]
```diff
@@ -23,6 +23,14 @@
...
 
 
+////////////////////////////////////////////////////////////////////////////////
+/// Destruct the engine, joining its thread.
+
+THttpWSEngine::~THttpWSEngine() {
+   if (fSendThrd.joinable())
+      fSendThrd.join();
+}
+
```
[the remark this answers is not in the document]
Comment (Axel-Naumann) [reply net/http/src/THttpWSEngine.cxx]: OK then you have a logic error somewhere. I'm hitting this in `draw_mt`. Please debug and suggest a different resolution. I'd prefer to still merge this as it at least fixes the test for now.

File: tutorials/v7/draw_mt.cxx [hunk 2]
```diff
@@ -91,7 +91,7 @@ void draw_mt()
    std::thread thrd2(draw_canvas, "Second canvas", RColor::kBlue);
    std::thread thrd3(draw_canvas, "Third canvas", RColor::kGreen);
 
-   thrd1.detach();
-   thrd2.detach();
-   thrd3.detach();
+   thrd1.join();
+   thrd2.join();
+   thrd3.join();
 }
```
[the remark this answers is not in the document]
Comment (Axel-Naumann) [reply tutorials/v7/draw_mt.cxx]: If I (or the test suite) runs with `root -q` we call `terminate` which is ever worse than not having the prompt. We might need some thread registry for interactivity, where upon global destruction threads are `join`ed.

File: net/http/src/THttpWSEngine.cxx (hunk 1 above)
[the remark this answers is not in the document]
Comment (linev) [reply net/http/src/THttpWSEngine.cxx]: This is a side effect of draw_mt.cxx termination without correct closure of user threads. 
If we do thrd1/2/3.join(), cleanup should work. 
As I proposed - one can do it in batch mode.

File: tutorials/v7/draw_mt.cxx (hunk 2 above)
[the remark this answers is not in the document]
Comment (linev) [reply tutorials/v7/draw_mt.cxx]: Yes, we missed this discussion completely during workshop.

One need different utilities to handle "interactive" threads, where user code plus some service ROOT functions must be executed. But simple join() does not help - such threads need to have some global termination condition on which they should react

Simple solution for now - enable join() for batch mode

File: net/http/src/THttpWSEngine.cxx (hunk 1 above)
[the remark this answers is not in the document]
Comment (Axel-Naumann) [reply net/http/src/THttpWSEngine.cxx]: Please try, it didn't help for me.

File: tutorials/v7/draw_mt.cxx (hunk 2 above)
[the remark this answers is not in the document]
Comment (Axel-Naumann) [reply tutorials/v7/draw_mt.cxx]: We cannot crash interactively if people `.q` too early. Tutorials need to work, properly. It's also a really bad example to `detach()` a thread and then just "leave". A tutorial is user-facing code and must be shiny and correct, without hacks. We can get rid of the `join()`s once we have the thread registration mechanism in ROOT.

File: net/http/src/THttpWSEngine.cxx (hunk 1 above)
[the remark this answers is not in the document]
Comment (linev) [reply net/http/src/THttpWSEngine.cxx]: > Please try, it didn't help for me.

Probably, after #2603 it will be better

File: tutorials/v7/draw_mt.cxx (hunk 2 above)
[the remark this answers is not in the document]
Comment (linev) [reply tutorials/v7/draw_mt.cxx]: > We cannot crash interactively if people .q too early. Tutorials need to work, properly. 

I agree. Since we do not have safe method to stop such threads - let use join.
I can add relevant parts of these changes into #2603 - anyway I need to adjust it now.

**Comment by linev** [comment]
More than 60 comments, 10 new commits :1st_place_medal: 
@Axel-Naumann killing me :)

I fix most of comments. Several big "stones" I left for next PRs:

1. use of `std::list<WebCommand>` in canvas painter. For me it really makes sense while here FIFO is used. And `queue` not very nice when I need to iterate over all items
2. Use of `collection::erase` methods in `WebWindow::CheckInactiveConnections` or `TWebWindow::CheckWebKeys`. @Axel-Naumann approach is very inefficient, while makes 10000 copies of same vector again and again without any need. I want to keep `erase`
3. C++ handle in WebWindowManager, which should be assigned with each started browser process. 
Fully agree, next PRs
4. WebWindowManager::Show() too big. Lets keep it for a while - it may change a lot, especially when solving point 3

**Comment by Axel-Naumann** [comment]
Was this mean to go to https://github.com/root-project/root/pull/2603 ? 'Cause this one is nice and simple.

Look, you get *way* fewer comments from me if you don't create PRs with 1400 lines changed in 32 files! You asked for it :-)

Much of these comments are also style/pattern comments; I hope that we can move to applying them in future code even more than seeing it fixed here. But yes, apologies for killing you ;-)

**Comment by linev** [comment]
> But yes, apologies for killing you ;-)

Yes, it was for #2603 
I hope now you can merge it to have good starting point for next steps.
Actually, one should rename these classes to be RWebWindow and RWebWindowManager.
Plus there are several inconsistencies in civetweb. First PR is #2584, but not last.

**Comment by Axel-Naumann** [comment]
Changed picked up by other PRs; lib loading extracted into PR #2676
